skimage/morphology/_max_tree.pyx

```diff
@@ -0,0 +1,459 @@
+                           ):
+    """Builds a max-tree.
+
+    Parameters
+    ----------
+
+    image : array of arbitrary type
+            The flattened image pixels.
+    mask : array of int
+        An array of the same shape as `image` where each pixel contains a
+        nonzero value if it is to be considered for the filtering.
+        NOTE: it is *essential* that the border pixels (those
+        with neighbors falling outside the volume) are all set to zero, or
+        segfaults could occur.
+    structure : array of int
+        A list of coordinate offsets to compute the raveled coordinates of each
+        neighbor from the raveled coordinates of the current pixel.
+    parent : array of int
+        First output: image of the same shape as the input image. The value
+        at each pixel is the parent index of this pixel in the max-tree
+        reprentation.
+    tree_order : array of int
+        Second output: list of length = number of pixels. Each element
+        corresponds to one pixel index in the image. It encodes the order
+        of elements in the tree: a parent of a pixel always comes before
+        the element itself. More formally: i < j implies that j cannot be
+        the parent of i.
+    """
+    cdef DTYPE_UINT64_t number_of_pixels = len(image)
+    cdef DTYPE_UINT64_t number_of_dimensions = len(shape)
+
+    cdef DTYPE_INT64_t i = 0
+    cdef DTYPE_INT64_t p = 0
+    cdef DTYPE_INT64_t root = 0
+    cdef DTYPE_INT64_t index = 0
+
+    cdef Py_ssize_t nneighbors = structure.shape[0]
+
+    cdef DTYPE_INT64_t[::1] zpar = parent.copy()
+
+    cdef np.ndarray[DTYPE_INT32_t, ndim = 2] points = offsets_to_points(structure, shape)
+
+    # initialization of the image parent.
+    for i in range(number_of_pixels):
+        parent[i] = -1
+        zpar[i] = -1
+
+    # traverse the array in reversed order (from highest value to lowest value)
+    for p in sorted_indices[::-1]:
+        parent[p] = p
+        zpar[p] = p
+
+        for i in range(nneighbors):
+
+            # get the ravelled index of the neighbor
+            index = p + structure[i]
+
+            if not mask[p]:
+                # in this case, p is at the border of the image.
+                # some neighbor point is not valid.
+                if not _is_valid_coordinate(p, points[i], shape):
+                    # neighbor is not in the image.
+                    continue
+
+            if parent[index] < 0:
+                # in this case the parent is not yet set: we ignore
+                continue
+
+            root = find_root(zpar, index)
+
+            if root != p:
+                zpar[root] = p
+                parent[root] = p
+
+    # In a canonized max-tree, each parent is a canonical pixel,
+    # i.e. for each connected component at a level l, all pixels point
+    # to the same representative which in turn points to the representative
+    # pixel at the next level.
+    canonize(image, parent, sorted_indices)
```

Comment:
Ah, and this explains why the max-tree image is so messy. Can we use a non-canonized version for the example? What are the performance implications of canonization? ie is it essential to canonize or only for some applications?

Reply:
Well, you had the right intuition : for an image without identical values all pixels are canonical. So, this is easy to do. Bad news is that it remains messy.

Reply:
I'm very slowly getting the idea of max-trees. When I first heard about them I thought it a parent could only be a neighbouring pixel, but that's not right, is it?

What is the property of an image that could make the tree spatially clean? If you had two increasing bright peaks, would you get the tree clean except for a node joining the two peaks?

Reply:
OK, I guess I got the right representation now (see below) to make all of this clear.